### ISSUE TITLE
refactor barcharts to be driven from scigraph

### ISSUE THREAD

**Issue by nlwashington** [monarch-initiative/monarch-legacy#681]
we need to refactor the datagraph/barcharts to be driven off of data from scigraph.

this could be done either directly from scigraph calls, or via golr views.

currently the four relevant barcharts are:
*phenotypes x annotation count (x species)  <-- hardest, but biggest reward
*disease x gene association count
*disease x phenotype annotation count
*phenotype x genotype (annotation) count

but an easy first one might be to build a new simple one as a foundation that is:
*phenotype x annotation count (for a single disease), like we've presented in talks (this one for [Schwartz-Jampel syndrome](http://monarchinitiative.org/disease/OMIM:255800):
![screen shot 2015-04-13 at 11 18 11 am](https://cloud.githubusercontent.com/assets/5124411/7122223/dcaad4b2-e1ce-11e4-8821-f7a2fee4cedf.png)

this is an "easy" one that would allow us to experiment with issues of drilling down in the ontology, and we could show these on each disease page.

**Comment by nlwashington** [comment]
@ccondit to think about cypher queries / Glor config possibilities.

**Comment by ccondit** [comment]
So a starting point would be a Golr query like this:

```
http://geoffrey.crbs.ucsd.edu:8080/solr/golr/select/?q=%28object_closure%3AHP%5C%3A0003011+OR+object_closure%3AHP%5C%3A0001939+OR+object_closure%3AHP%5C%3A0001197+OR+object_closure%3AHP%5C%3A0000769+OR+object_closure%3AHP%5C%3A0001871+OR+object_closure%3AHP%5C%3A0000478+OR+object_closure%3AHP%5C%3A0002664+OR+object_closure%3AHP%5C%3A0045027+OR+object_closure%3AHP%5C%3A0040064+OR+object_closure%3AHP%5C%3A0001626+OR+object_closure%3AHP%5C%3A0000707+OR+object_closure%3AHP%5C%3A0001608+OR+object_closure%3AHP%5C%3A0002715+OR+object_closure%3AHP%5C%3A0003549+OR+object_closure%3AHP%5C%3A0001574+OR+object_closure%3AHP%5C%3A0000119+OR+object_closure%3AHP%5C%3A0001438+OR+object_closure%3AHP%5C%3A0000598+OR+object_closure%3AHP%5C%3A0001507+OR+object_closure%3AHP%5C%3A0000118+OR+object_closure%3AHP%5C%3A0000924+OR+object_closure%3AHP%5C%3A0002086+OR+object_closure%3AHP%5C%3A0000818+OR+object_closure%3AHP%5C%3A0000152%29%0D%0AAND+subject%3AOMIM%5C%3A255800&version=2.2&start=0&rows=10&indent=on
```

Or more generally:
`(object_closure:phenotypeChildId1 OR objectClosure:phenotypeChildIdN) AND subject:diseaseId`

The only trick here is going to be getting the facets right. You may have to iterate through the closures and check for the presence of a phenotypeChild and group in the application.

**Comment by ccondit** [comment]
Actually even better yet append `&facet.field=object_closure&facet=true` to the URL in my previous comment and you'll get the facets. You'll just need to filter them out to the ones that were originally input.

**Comment by ccondit** [comment]
Here's a combined, improved version:

```
http://geoffrey.crbs.ucsd.edu:8080/solr/golr/select/?q=%28object_closure%3AHP%5C%3A0003011+OR+object_closure%3AHP%5C%3A0001939+OR+object_closure%3AHP%5C%3A0001197+OR+object_closure%3AHP%5C%3A0000769+OR+object_closure%3AHP%5C%3A0001871+OR+object_closure%3AHP%5C%3A0000478+OR+object_closure%3AHP%5C%3A0002664+OR+object_closure%3AHP%5C%3A0045027+OR+object_closure%3AHP%5C%3A0040064+OR+object_closure%3AHP%5C%3A0001626+OR+object_closure%3AHP%5C%3A0000707+OR+object_closure%3AHP%5C%3A0001608+OR+object_closure%3AHP%5C%3A0002715+OR+object_closure%3AHP%5C%3A0003549+OR+object_closure%3AHP%5C%3A0001574+OR+object_closure%3AHP%5C%3A0000119+OR+object_closure%3AHP%5C%3A0001438+OR+object_closure%3AHP%5C%3A0000598+OR+object_closure%3AHP%5C%3A0001507+OR+object_closure%3AHP%5C%3A0000118+OR+object_closure%3AHP%5C%3A0000924+OR+object_closure%3AHP%5C%3A0002086+OR+object_closure%3AHP%5C%3A0000818+OR+object_closure%3AHP%5C%3A0000152%29%0D%0AAND+subject%3AOMIM%5C%3A255800&version=2.2&start=0&rows=0&indent=on&facet.field=object_closure&facet=true
```

**Comment by kltm** [comment]
@ccondit it might be easier/more robust to use one of the managers and response handlers rather than raw URLs and parsing the response:

 https://kltm.github.io/bbop-js/docs/files/golr/manager-js.html
 https://kltm.github.io/bbop-js/docs/files/golr/response-js.html

These are being used elsewhere by @kshefchek already.

**Comment by ccondit** [comment]
@kltm - absolutely yes.

SciGraph can also get the new list of candidate object ids as the client drills down:

```
http://geoffrey.crbs.ucsd.edu:9000/scigraph/graph/neighbors/HP:0000118?depth=1&blankNodes=false&direction=in&relationshipType=subClassOf
```

if you need that.

**Comment by nlwashington** [comment]
(thinking out loud here) i think this is fairly straightforward to generate a single disease on-the-fly.  given a disease of interest identifier:
1.  we will draw the initial chart with some configured set of top-level nodes (i'll call a category here).  for phenotypes, it's probably the children of HP:0000118 (phenotype).  these categories can be fetched from SciGraph dynamically, if desired.
2.  given the set of category ids and the disease of interest id, construct a Golr query as above, and make the request.  since the annotations in scigraph are constructed "disease has_phenotype phenotype", it fetches all items with with the category ids in the object_closure.  if we have already fetched the associations (as on our new disease-golr pages), then we can just skip this step.
3.  the list of results can be iterated over to bin the specific phenotype into the relevant category to get a summary count by category.  first could be to build a flat hashmap of  `{ phenotype id : count }`, where the count is simply the count of occurrences of each item over all the object_closures.  alternatively, we turn the object_closure list inside out, so that instead of a simple count, we store the array of specific phenotypes that map to the more-general phenotypes.
4.  the chart is drawn by looking up the category ids in the hash for counts.  
5.  in order to drill down, a new SciGraph query can be constructed with the selected category to fetch its children and get new categories, as in step 1.  but now, we can use the stored hashmap of phenotype counts to populate and redraw the chart as in step 4.  

this will work fine for leaf nodes.  for intermediate nodes, the subject of the annotation is not unique (>1 disease).  we will want to consider offering the user options for observing only unique phenotypes for all diseases, or redundant phenotypes.  this would be interesting because it would change the distribution...maybe via radiobutton. 

doing this using DOID:4 would give us the disease-to-phenotype overview for all diseases, as we had before.  we should test this out to see if it is efficient to do the root of the disease ontology on the fly, or if it needs to be precomputed and stored.  (there's probably a lot to iterate through, and i suspect it to be slow.)

**Comment by cmungall** [comment]
I'm changing the milestone as we can't be drawing data from federation and SG/golr in the next release. Putting under phenotype as it's primarily phenotype association counts

**Comment by kshefchek** [comment]
Some notes and a follow up from talking with @kltm:

1) We need to query GOlr once per class since we need facet counts per species, and we cannot get facet counts for closures on a per species basis.  In the future this will be updated for a "pivot table" view of the data so we can get this with fewer calls.
2) The golr jquery manager has a run_batch function that will launch jobs serially, however this would be a bit slower than
3) We can launch ajax requests in parallel, perhaps using promises, but we run into issues with number of classes per view, as many sibling groups contain 15+ classes, and browsers limiting ajax requests, investigating now.

**Comment by kshefchek** [comment]
This refactor is almost finished, the new graph is visible on the phenotype landing page and will work once geoffrey golr is back online.  From there it will be easy to add various graphs to other pages as the query configuration is set like this: 
https://github.com/monarch-initiative/monarch-app/blob/master/js/graph-config.js#L91
And we can dynamically add additional filters if we want to narrow in on a specific id (say on an overview page)

Biggest issues right now are displaying large groups of classes (some groups have 150+ classes).  I need to add some code that will redraw the svg area with a new height in these cases.  Another issue is clicking on leaf nodes results in the graph transitioning to an empty chart.  I need to add a message in these cases and prevent the graph from attempting to transition.

**Comment by nlwashington** [comment]
can the height dynamically size larger to a certain max?  or perhaps enable scrolling if we add a max height?

**Comment by kshefchek** [comment]
I think that is doable.  Scrolling should default to the browsers scrolling mechanism as we're not using a viewport.

**Comment by kshefchek** [comment]
I've update all of the barcharts to be driven from SciGraph and GOlr and fixed the issues above.  Please reopen if needed or break out any issues into a new ticket.
